dansjack/ad-470-group2
# Flask Web Server

# Description 📄 
Closes #9
Closes #10 


This PR adds a Flask server that includes a question answering bot and some info about our project. The connection between 
<img width="800" alt="image" src="https://user-images.githubusercontent.com/30481844/170401760-44c3096d-1c03-4842-a3b0-3311c6fb178d.png">

## Project Details 💅 
**Styles**: [Tailwindcss](https://tailwindcss.com/)
**DB**: Sqlite3. The database will only save the chat log for the current active instance. If you redeploy, the db will by wiped.
**Default Model**: [HuggingFace](https://huggingface.co/deepset/bert-base-cased-squad2?context=My+name+is+Sarah+and+I+live+in+London&question=Where+do+I+live%3F) BERT model pretrained on the SQuAD2.0 data set
**Model Context Provider**: [Wikipedia API](https://wikipedia.readthedocs.io/en/latest/code.html#api)
**Model Reader**: The `DocumentReader` class comes from a blog by **Cloudera Fast Forward Labs**,  [Building a QA System with BERT on Wikipedia](https://qa.fastforwardlabs.com/pytorch/hugging%20face/wikipedia/bert/transformers/2020/05/19/Getting_Started_with_QA.html#QA-on-Wikipedia-pages). This class is very important because it is used to tokenize the question and context, and then retrieve an answer
**Deployment:** [Docker](https://www.docker.com/) (local), and [Google Cloud Run](https://cloud.google.com/run) (cloud)

## Question Answering options ❓ 
There are two options:
* Model type: select the model you wish the bot to answer with
    * pretrained (default): A model already trained with the SQuAD2.0 dataset. See **Default Model** above for more info
    * custom: an untrained model (currently a placeholder for our model, the answers returned will be nonsense)
* Answer type: select the format of the answer returned by the bot
    * first (default): return the very first match from the Wikipedia page
    * all: return all matches from the Wikipedia page

## How it works 📖 
When you ask the bot a question, it will:
1. Search for the most relevant page on Wikipedia via the Wikipedia API
    * If no page could be found, the bot responds with: "I checked Wikipedia and couldn't find a good page for your question. Please try another question."
2. The `DocumentReader` tokenizes the question and the text from the Wikipedia page in step 1
3. The `DocumentReader` then parses and returns an answer.
    * If no suitable answer could be found on the page, the bot responds with: "You stumped me! I don't know the answer."
4. The answer is returned by the `DocumentReader` and displayed to the user

_Note: There are a few edge cases where the Wikipedia API returns an error or a value the `DocumentReader` isn't expecting. In those cases, the bot will respond with either:_
* "Oops, I had a problem reading the suggested page for your question. Can you try asking something else?"
* "Hmm, Wikipedia gave me some data I didn't know how to handle. Can you try asking something else?"

# Testing 🧪 
1. Try asking a few questions of the qa bot at https://server-ejyiaysa3q-uc.a.run.app/
2. Try setting up the local server using the instructions found in the README (see comment below)
